src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,452 @@
+use std::{
+    cell::UnsafeCell, mem::transmute, sync::{atomic::{AtomicU64, AtomicUsize}, Arc}, time::Duration, u8
+};
+
+use bytes::BytesMut;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{self, TcpStream},
+};
+use tracing::debug;
+
+use crate::{
+    common::TimeoutOptions, error::RpcError, message::{ReqType, RespType}, packet::{
+        Decode, Encode, Packet, PacketStatus, PacketsKeeper, ReqHeader, RespHeader, REQ_HEADER_SIZE
+    }, read_exact_timeout, write_all_timeout
+};
+
+/// TODO: combine RpcClientConnectionInner and RpcClientConnection
+struct RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The TCP stream for the connection.
+    stream: UnsafeCell<net::TcpStream>,
+    /// Options for the timeout of the connection
+    timeout_options: TimeoutOptions,
+    /// Stream auto increment sequence number, used to mark the request and response
+    seq: AtomicU64,
+    /// Received keep alive seq number, will mark the valid keep alive response.
+    /// If the keep alive response is not received in 100 times, the connection will be closed,
+    /// If we receive other response, we will update the received_keepalive_seq too, just treat it as a keep alive response.
+    received_keepalive_seq: AtomicU64,
+    /// Send packet task
+    packets_keeper: UnsafeCell<PacketsKeeper<P>>,
+    /// Response buffer, try to reuse same buffer to reduce memory allocation
+    /// In case of the response is too large, we need to consider the buffer size
+    /// Init size is 4MB
+    /// Besides, we want to reduce memory copy, and read/write the response to the same data buffer
+    resp_buf: UnsafeCell<BytesMut>,
+}
+
+impl<P> RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static
+{
+    pub fn new(stream: net::TcpStream, timeout_options: TimeoutOptions) -> Self {
+        Self {
+            stream: UnsafeCell::new(stream),
+            timeout_options: timeout_options.clone(),
+            seq: AtomicU64::new(0),
+            received_keepalive_seq: AtomicU64::new(0),
+            packets_keeper: UnsafeCell::new(PacketsKeeper::new(timeout_options.clone().idle_timeout.as_secs())),
+            resp_buf: UnsafeCell::new(BytesMut::with_capacity(8 * 1024 * 1024)),
+        }
+    }
+
+    /// Recv request header from the stream
+    pub async fn recv_header(&self) -> Result<RespHeader, RpcError<String>> {
+        // Try to read to buffer
+        match self.recv_len(REQ_HEADER_SIZE).await {
```

Comment:
recv header at client-side should not use read_exact_timeout, because it depends on whether there's packet sent to server.  (meanswhile  recv body should use read timeout)

Reply:
So we need to block the read here?

Reply:
Yes, block until Client is dropped or stream error. write a future to check the closed channel

Reply:
If we keep blocking here, it may affect the scheduling of tokio on the client side and cause other tasks not to be called normally.

Recv is similar to try, and will be completed by loop {}.

In addition, the timeout for recv is long enough for the current length of the packet to be read.

Reply:
OK, I added a StreamReaderFuture to check client and stream is closed.

Reply:
`RecvLenFuture`